### PR TITLE
feat!: add versioning for pubsub topics in fraud, header packages

### PR DESCRIPTION
**Pull request by vgonkivs**
## Overview
Resolves https://github.com/celestiaorg/celestia-node/issues/1455, https://github.com/celestiaorg/celestia-node/issues/1454, https://github.com/celestiaorg/celestia-node/issues/1453

PLEASE NOTE: this PR is a protocol breaking since it introduces changes in protocol namings.

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: fraud/requester.go [hunk 1]
```diff
@@ -20,11 +21,12 @@ const (
 
 func (f *ProofService) requestProofs(
 	ctx context.Context,
+	id protocol.ID,
```
Comment (distractedm1nd) [inline fraud/requester.go:24]: One protocol ID for many proof types - we discussed in sync maybe having a default protocol ID (without a proof type) that is valid for all types

**Review by distractedm1nd — CHANGES_REQUESTED** [review]

File: fraud/requester.go (hunk 1 above)
Comment (vgonkivs) [reply fraud/requester.go:24]: `id` does not contain proofType. So, we will request all available fraud proofs.

File: libs/header/p2p/subscriber.go [hunk 2]
```diff
@@ -13,30 +13,38 @@ import (
...
 	pubsub *pubsub.PubSub
 	topic  *pubsub.Topic
 	msgID  pubsub.MsgIdFunction
 }
 
 // NewSubscriber returns a Subscriber that manages the header Module's
 // relationship with the "header-sub" gossipsub topic.
-func NewSubscriber[H header.Header](ps *pubsub.PubSub, msgID pubsub.MsgIdFunction) *Subscriber[H] {
+func NewSubscriber[H header.Header](
+	protocolSuffix string,
```
Comment (Wondertan) [inline libs/header/p2p/subscriber.go:26]: nit: last param

**Review by Wondertan — APPROVED** [review]
Lgtm 
Does anyone remember why we went with suffixes and not prefixes? /celestia/arabica/header-ex/000 feels better than header-ex/000/celestia/arabica
Usually you start with higher categories and then go deeper.

**Comment by vgonkivs** [comment]
I think we haven't discussed it yet. But, think it's better to have a constant `/celestia/<network>/header-ex/` + incrementing suffix

File: core/listener_test.go [hunk 3]
```diff
@@ -104,7 +103,7 @@ func createListener(
 	ps *pubsub.PubSub,
 	edsSub *shrexsub.PubSub,
 ) *Listener {
-	p2pSub := p2p.NewSubscriber[*header.ExtendedHeader](ps, header.MsgID)
+	p2pSub := p2p.NewSubscriber[*header.ExtendedHeader](ps, header.MsgID, "private")
```
Comment (renaynay) [inline core/listener_test.go:106]: here too

File: libs/header/p2p/exchange_test.go [hunk 4]
```diff
@@ -94,19 +94,24 @@ func TestExchange_RequestFullRangeHeaders(t *testing.T) {
 	hosts := createMocknet(t, 5)
 	totalAmount := 80
 	store := headerMock.NewStore[*test.DummyHeader](t, test.NewTestSuite(t), totalAmount)
-	protocolSuffix := "private"
 	connGater, err := conngater.NewBasicConnectionGater(sync.MutexWrap(datastore.NewMapDatastore()))
 	require.NoError(t, err)
 	// create new exchange
-	exchange, err := NewExchange[*test.DummyHeader](hosts[len(hosts)-1], []peer.ID{}, protocolSuffix, connGater)
+	exchange, err := NewExchange[*test.DummyHeader](hosts[len(hosts)-1], []peer.ID{}, connGater,
+		WithProtocolSuffix[ClientParameters]("private"),
```
Comment (renaynay) [inline libs/header/p2p/exchange_test.go:101]: here too re `p2p.Private`

File: libs/header/p2p/exchange_test.go [hunk 5]
```diff
@@ -94,19 +94,24 @@ func TestExchange_RequestFullRangeHeaders(t *testing.T) {
...
 	exchange.ctx, exchange.cancel = context.WithCancel(context.Background())
 	t.Cleanup(exchange.cancel)
 	// amount of servers is len(hosts)-1 because one peer acts as a client
 	servers := make([]*ExchangeServer[*test.DummyHeader], len(hosts)-1)
 	for index := range servers {
-		servers[index], err = NewExchangeServer[*test.DummyHeader](hosts[index], store, protocolSuffix)
+		servers[index], err = NewExchangeServer[*test.DummyHeader](
+			hosts[index],
+			store,
+			WithProtocolSuffix[ServerParameters]("private"),
```
Comment (renaynay) [inline libs/header/p2p/exchange_test.go:113]: same here

File: libs/header/p2p/subscriber.go [hunk 6]
```diff
@@ -13,30 +13,38 @@ import (
 // Subscriber manages the lifecycle and relationship of header Module
 // with the "header-sub" gossipsub topic.
 type Subscriber[H header.Header] struct {
+	pubSubTopicID string
```
Comment (renaynay) [inline libs/header/p2p/subscriber.go:16]: nit but 
```suggestion
	pubsubTopicID string
```

File: share/p2p/shrexsub/pubsub.go [hunk 7]
```diff
@@ -17,7 +17,7 @@ var log = logging.Logger("shrex-sub")
 
 // pubSubTopic hardcodes the name of the EDS floodsub topic with the provided suffix.
 func pubSubTopic(suffix string) string {
-	return "eds-sub/v0.0.1/" + suffix
+	return fmt.Sprintf("/eds-sub/v0.0.1/%s", suffix)
```
Comment (renaynay) [inline share/p2p/shrexsub/pubsub.go:20]: are you sure this won't have a double `//` prepended?

**Review by renaynay — COMMENTED** [review]
some nits 

can you change from using `"private"` to `p2p.Private` in tests pls 🙏🏻

File: share/p2p/shrexsub/pubsub.go (hunk 7 above)
Comment (vgonkivs) [reply share/p2p/shrexsub/pubsub.go:20]: no. it should not

File: fraud/helpers.go [hunk 8]
```diff
@@ -9,17 +9,17 @@ import (
 	"github.com/libp2p/go-libp2p/core/protocol"
 )
 
-func pubSubTopicID(fraudType, protocolSuffix string) string {
-	return fmt.Sprintf("/fraud-sub/%s/v0.0.1/%s", fraudType, protocolSuffix)
+func pubsubTopicID(fraudType, protocolSuffix string) string {
+	return fmt.Sprintf("/fraud-sub/%s/%s/v0.0.1", protocolSuffix, fraudType)
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/helpers.go:13]: Ok. I agree. Was thinking between these two cases and chose mine.

[the remark this answers is not in the document]
Comment (vgonkivs) [reply fraud/helpers.go:13]: discussed internally. I agree

**Review by distractedm1nd — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Pull request merged**